### PR TITLE
V7 - useFieldArray type issue

### PR DESCRIPTION
**Pull request by bluebill1049**
Huge thanks go to https://github.com/phpnode 's help on type definition. 

- fix useFieldArray with register index with variable method
- improve watch API type
- improve getValues API type and functionality 
- getValues() will return and shallow merge with defaultValues

### CODE REVIEW

File: src/types/utils.ts [hunk 1]
```diff
@@ -66,119 +67,156 @@ export type IsFlatObject<T extends object> = Extract<
...
+    : ArrayKey | `${ArrayKey}.${FieldPath<E>}`
+  : ValueOf<
+      {
+        [K in keyof Root]: IsPrimitive<Root[K]> extends true
+          ? K & string
+          : (K & string) | `${K & string}.${FieldPath<Root[K]>}`;
+      }
+    >;
+
+export type FieldPathValue<
```
Comment (bluebill1049) [inline src/types/utils.ts:209]: issue with `FieldPathValue` not getting the correct type

**Comment by bluebill1049** [comment]
nested field array still a problem.

File: src/types/utils.ts (hunk 1 above)
Comment (kotarella1110) [reply src/types/utils.ts:209]: @bluebill1049 I will fix this 👍

Comment (bluebill1049) [reply src/types/utils.ts:209]: I think I got it working last night and just committed now, please double check.

Comment (bluebill1049) [reply src/types/utils.ts:209]: nvm, it's not working...
Screen Shot 2021-01-08 at 12 38 15 pm

**Comment by bluebill1049** [comment]
let's merge this in and the `register` branch then, we can branch off from the main branch.

**Pull request merged**